src/backend/distributed/commands/multi_copy.c

```diff
@@ -186,6 +190,187 @@ static void CopyAttributeOutText(CopyOutState outputState, char *string);
 static inline void CopyFlushOutput(CopyOutState outputState, char *start, char *pointer);
 
 
+bool
+WorkerCopy(CopyStmt *copyStatement)
+{
+	ListCell *optionCell = NULL;
+
+	/* Extract options from the statement node tree */
+	foreach(optionCell, copyStatement->options)
+	{
+		DefElem *defel = (DefElem *) lfirst(optionCell);
+
+		if (strncmp(defel->defname, "master_hostname", NAMEDATALEN) == 0)
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+NodeAddress *
+MasterNodeAddress(CopyStmt *copyStatement)
+{
+	NodeAddress *masterNodeAddress = (NodeAddress *) palloc0(sizeof(NodeAddress));
+	ListCell *optionCell = NULL;
+
+	/* set default port */
+	masterNodeAddress->nodePort = 5432;
+
+	/* Extract options from the statement node tree */
+	foreach(optionCell, copyStatement->options)
+	{
+		DefElem *defel = (DefElem *) lfirst(optionCell);
+
+		if (strncmp(defel->defname, "master_hostname", NAMEDATALEN) == 0)
+		{
+			masterNodeAddress->nodeName = defGetString(defel);
+		}
+		else if (strncmp(defel->defname, "master_port", NAMEDATALEN) == 0)
+		{
+			masterNodeAddress->nodePort = defGetInt32(defel);
+		}
+	}
+
+	return masterNodeAddress;
+}
+
+
+/*
+ * RemoveMasterOption removes master node related copy options from the option
+ * list of the copy statement.
+ */
+static void
+RemoveMasterOption(CopyStmt *copyStatement)
+{
+	List *newOptionList = NIL;
+	ListCell *optionCell = NULL;
+
+	/* Extract options from the statement node tree */
+	foreach(optionCell, copyStatement->options)
+	{
+		DefElem *option = (DefElem *) lfirst(optionCell);
+
+		if ((strncmp(option->defname, "master_hostname", NAMEDATALEN) == 0) ||
+			(strncmp(option->defname, "master_port", NAMEDATALEN) == 0))
+		{
+			continue;
+		}
+
+		newOptionList = lappend(newOptionList, option);
+	}
+
+	copyStatement->options = newOptionList;
+}
+
+
+static char
+RemotePartitionMethod(PGconn *connection, char *relationName)
```

Comment:
> It's a bit odd that the UDF is called partition_type, but elsewhere we call it partition method.

partition_type udf was already there and it uses PartitionMethod() function. Somewhere we call it as "partition type" and somewhere as "partition method". Ideally, we should unify this but probably not in this PR.

What do you think?